flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,207 @@
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateBufferPoolSize(inputGate1));
+
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateBufferPoolSize(inputGate2));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateBufferPoolSize(inputGate2));
+
+			List<Buffer> exclusiveBuffers = new ArrayList<>();
+
+			// test for exclusive buffer used
+			exclusiveBuffers.add(remoteInputChannel1.requestBuffer());
+			assertEquals(1, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+
+			for (int i = 0; i < buffersPerChannel; i++){
+				Buffer buffer = remoteInputChannel1.requestBuffer();
+				if (buffer != null) {
+					exclusiveBuffers.add(buffer);
+				} else {
+					break;
+				}
+			}
+			assertEquals(buffersPerChannel, exclusiveBuffers.size());
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals(1 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals(buffersPerChannel / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
+
+			for (int i = 0; i < buffersPerChannel; i++){
+				Buffer buffer = remoteInputChannel2.requestBuffer();
+				if (buffer != null) {
+					exclusiveBuffers.add(buffer);
+				} else {
+					break;
+				}
+			}
+			assertEquals(buffersPerChannel * 2, exclusiveBuffers.size());
+			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals(2 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals(buffersPerChannel * 2 / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
+
+			// test for floating buffer used
+			List<Buffer> floatingBuffers = new ArrayList<>();
+
+			remoteInputChannel1.requestSubpartition(0);
+			remoteInputChannel1.onSenderBacklog(3);
+			assertEquals(buffersPerChannel + 3, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+
+			for (int i = 0; i < buffersPerChannel + 3; i++) {
+				Buffer buffer = remoteInputChannel1.requestBuffer();
+				if (buffer != null) {
+					floatingBuffers.add(buffer);
+				} else {
+					break;
+				}
+			}
+			assertEquals(0, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals(2 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals((buffersPerChannel * 2 + buffersPerChannel + 3) / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
+
+			List<MemorySegment> segments = new ArrayList<>();
+			for (Buffer buffer : exclusiveBuffers) {
+				segments.add(buffer.getMemorySegment());
+			}
+			network.getNetworkBufferPool().recycleMemorySegments(segments);
```

Comment:
Couple of questions:
1. Shouldn't this be a part of `finally`?
2. Couldn't you call `buffer.recycle()`?
3. Maybe it would be better to run whole test with a `ClosableRegistry` that you would be populating as you create more objects? For example input gates there once you create them, or register `() -> { buffer.recycle() }` every time you just requested a buffer? It would make closing/releasing resource more reliable and more fine grained.